### PR TITLE
AVX512VL enabling

### PR DESCRIPTION
**Pull request by deeptiag1**

### CODE REVIEW

File: src/generator_gemm_common.c [hunk 1]
```diff
@@ -1586,7 +1974,7 @@ void libxsmm_generator_gemm_initialize_avx512_mask( libxsmm_generated_code*
       i_gp_reg_tmp,
       l_mask );
 
-  if ( ( io_generated_code->arch >= LIBXSMM_X86_AVX512 ) && ( io_generated_code->arch <= LIBXSMM_X86_ALLFEAT ) ) {
+  if ( ( io_generated_code->arch >= LIBXSMM_X86_AVX512_VL256 ) && ( io_generated_code->arch <= LIBXSMM_X86_ALLFEAT ) ) {
```
Comment (alheinecke) [inline src/generator_gemm_common.c:1977]: Hmm.... There are a lot diffs here which make no sense in the file... Is this a git displaying issue?

File: src/generator_gemm_sse_avx_avx2_avx512.c [hunk 2]
```diff
@@ -111,7 +111,8 @@ LIBXSMM_API_INTERN void libxsmm_generator_gemm_sse_avx_avx2_avx512_kernel( libxs
   libxsmm_generator_gemm_init_micro_kernel_config_fullvector( &l_micro_kernel_config, io_generated_code->arch, i_xgemm_desc, 0 );
 
   /* block according to the number of available registers or given limits */
-  if ( i_xgemm_desc->n == 7 && io_generated_code->arch >= LIBXSMM_X86_AVX512_CORE && io_generated_code->arch <= LIBXSMM_X86_ALLFEAT ) {
+  if ( i_xgemm_desc->n == 7 && io_generated_code->arch > LIBXSMM_X86_AVX512_VL256 && io_generated_code->arch != LIBXSMM_X86_AVX512_VL256_CLX
+        && io_generated_code->arch != LIBXSMM_X86_AVX512_VL256_CPX && io_generated_code->arch <= LIBXSMM_X86_ALLFEAT ) {
```
Comment (alheinecke) [inline src/generator_gemm_sse_avx_avx2_avx512.c:115]: This change seems to be a no-op

File: src/generator_mateltwise_reduce_avx_avx512.c [hunk 3]
```diff
@@ -3335,15 +3335,15 @@ void libxsmm_generator_opreduce_vecs_index_avx512_microkernel( libxsmm_generated
       reduceop_instr = LIBXSMM_X86_INSTR_VRANGEPS;
       reduceop_imm = 5;
       argop_cmp_imm = 6;
-      if (io_generated_code->arch < LIBXSMM_X86_AVX512_CORE) {
+      if (io_generated_code->arch < LIBXSMM_X86_AVX512_VL256) {
```
Comment (alheinecke) [inline src/generator_mateltwise_reduce_avx_avx512.c:3338]: KNL/KNM not handled correctly

File: src/generator_mateltwise_reduce_avx_avx512.c [hunk 4]
```diff
@@ -3335,15 +3335,15 @@ void libxsmm_generator_opreduce_vecs_index_avx512_microkernel( libxsmm_generated
...
         reduceop_instr = LIBXSMM_X86_INSTR_VMAXPS;
         reduceop_imm = 0;
       }
     } else if ((i_mateltwise_desc->flags & LIBXSMM_MELTW_FLAG_OPREDUCE_VECS_REDOP_MIN) > 0) {
       reduceop_instr = LIBXSMM_X86_INSTR_VRANGEPS;
       reduceop_imm = 4;
       argop_cmp_imm = 9;
-      if (io_generated_code->arch < LIBXSMM_X86_AVX512_CORE) {
+      if (io_generated_code->arch < LIBXSMM_X86_AVX512_VL256) {
```
Comment (alheinecke) [inline src/generator_mateltwise_reduce_avx_avx512.c:3346]: KNL/KNM not handled correctly

File: src/libxsmm_dnn.c [hunk 5]
```diff
@@ -21,7 +21,12 @@
 #if defined(LIBXSMM_OFFLOAD_TARGET)
 # pragma offload_attribute(pop)
 #endif
-
+#define block64
+#if defined block64
+# define blocksize 64
+#else
+# define blocksize 32
+#endif
```
Comment (alheinecke) [inline src/libxsmm_dnn.c:29]: Minor issue.. all caps with LIBXSMM_ with prefix for defines would be great.

File: src/libxsmm_dnn_convolution_forward.c [hunk 6]
```diff
@@ -22,7 +22,7 @@ LIBXSMM_API_INTERN libxsmm_dnn_err_t libxsmm_dnn_convolve_st_fwd_custom_custom_i
 LIBXSMM_API_INTERN libxsmm_dnn_err_t libxsmm_dnn_convolve_st_fwd_custom_custom_i8_i8(libxsmm_dnn_layer* handle, int start_thread, int tid);
 
 
-LIBXSMM_API_INTERN LIBXSMM_INTRINSICS(LIBXSMM_X86_AVX512)
+LIBXSMM_API_INTERN LIBXSMM_INTRINSICS(LIBXSMM_X86_AVX512_VL256)
 libxsmm_dnn_err_t libxsmm_dnn_convolve_st_fwd_custom_custom_f32_f32(libxsmm_dnn_layer* handle, int start_thread, int tid)
```
Comment (alheinecke) [inline src/libxsmm_dnn_convolution_forward.c:26]: I'm not sure about the intrsinics tests... It's technically still AVX512... so I would not change this.

File: include/libxsmm_intrinsics_x86.h [hunk 7]
```diff
@@ -258,11 +258,21 @@
 #       else
 #         define LIBXSMM_ATTRIBUTE_TARGET_1006 LIBXSMM_ATTRIBUTE_TARGET_1005
 #       endif
-#       if (LIBXSMM_X86_AVX512 <= LIBXSMM_MAX_STATIC_TARGET_ARCH)
+#       if (LIBXSMM_X86_AVX512_VL256 <= LIBXSMM_MAX_STATIC_TARGET_ARCH)
```
Comment (alheinecke) [inline include/libxsmm_intrinsics_x86.h:261]: I would not add these intrinsics version, as we technically need full AVX512 support of the compiler, but later in convolutions this intrinsics are used as well. (see further down below).

**Review by alheinecke — CHANGES_REQUESTED** [review]

File: src/generator_mateltwise_reduce_avx_avx512.c (hunk 3 above)
Comment (deeptiag1) [reply src/generator_mateltwise_reduce_avx_avx512.c:3338]: This change was not intended for m8 Blocking. I will revert it back. Eltwise changes still needs to be tested!

File: src/generator_mateltwise_reduce_avx_avx512.c (hunk 4 above)
Comment (deeptiag1) [reply src/generator_mateltwise_reduce_avx_avx512.c:3346]: This change was not intended for m8 Blocking. I will revert it back. Eltwise changes still needs to be tested!

File: src/generator_gemm_common.c [hunk 8]
```diff
@@ -1079,15 +1281,15 @@ void libxsmm_generator_gemm_load_C( libxsmm_generated_code*             io_gener
       }
     /* pure int8 kernel */
     } else if ( ( (i_micro_kernel_config->instruction_set >= LIBXSMM_X86_AVX512_CORE) && (i_micro_kernel_config->instruction_set <= LIBXSMM_X86_ALLFEAT) ) &&
-                ( (LIBXSMM_GEMM_PRECISION_I8 == LIBXSMM_GETENUM_INP( i_xgemm_desc->datatype ) ) && (LIBXSMM_GEMM_PRECISION_I8 == LIBXSMM_GETENUM_OUT( i_xgemm_desc->datatype ) ) ) ) {
-      /* we need to up convert int8 to int32 */
+                ( (LIBXSMM_GEMM_PRECISION_BF16 == LIBXSMM_GETENUM_INP( i_xgemm_desc->datatype ) ) && (LIBXSMM_GEMM_PRECISION_BF16 == LIBXSMM_GETENUM_OUT( i_xgemm_desc->datatype ) ) ) ) {
+      /* we add when scaling during conversion to FP32 */
```
[the remark this answers is not in the document]
Comment (deeptiag1) [reply src/generator_gemm_common.c:1285]: The comment is misleading here. It is BF16 change.

**Pull request merged**